source/MaterialXGenShader/ShaderTranslator.h

```diff
@@ -0,0 +1,75 @@
+//
+// TM & (c) 2020 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#ifndef MATERIALX_SHADERTRANSLATOR_H
+#define MATERIALX_SHADERTRANSLATOR_H
+
+#include <MaterialXGenShader/Library.h>
+#include <MaterialXGenShader/Shader.h>
+#include <MaterialXGenShader/Util.h>
+
+namespace MaterialX
+{
+
+/// A shared pointer to a ShaderTranslator
+using ShaderTranslatorPtr = shared_ptr<class ShaderTranslator>;
+
+/// @class ShaderTranslator
+class ShaderTranslator
+{
+  public:
+    static ShaderTranslatorPtr create(ConstDocumentPtr doc)
+    {
+        return ShaderTranslatorPtr(new ShaderTranslator(doc));
+    }
+
+    /// Translates shaderRef to the destShader shading model
+    void translateShader(ShaderRefPtr shaderRef, string destShader);
+
+    /// Translates all the materials to the destShader shading model if translation exists.
+    static bool translateAllMaterials(DocumentPtr doc, string destShader);
+
+    /// Returns set of all the available potential translations
+    StringSet getAvailableTranslations(string start)
+    {
+        return _shadingTranslations[start];
+    }
+
+  protected:
+    ShaderTranslator(ConstDocumentPtr doc);
+
+    /// Reads shading translation nodes from the document
+    void loadShadingTranslations();
+
+    /// Connects translation node inputs from the original shaderRef
+    void connectToTranslationInputs(ShaderRefPtr shaderRef);
+
+    /// Copies translation nodegraph upstream node dependencies over to the working nodegraph.
+    /// Used when normals need to be baked in tangent space but shaderref expects normals to 
+    /// be in world space.
+    void insertOutputNgUpstreamElems(OutputPtr translatedOutput, OutputPtr ngOutput);
```

Comment:
One additional suggestion that didn't come through in my last review:

Following the logic of your documentation above, would ShaderTranslator::insertUpstreamDependencies be a clearer name for this method?